### PR TITLE
Performance Changes for Weak Event Pattern

### PR DESCRIPTION
**Pull request by prashelke**
Thank you for your contribution.
Before submitting this PR, please make sure:
- [x] PR description and commit message should describe the changes done in this PR
- [x] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [x] Latest Code from master or specific release branch is merged to your branch
- [x] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Help Libray document is updated accordingly for Feature changes.

### CODE REVIEW

File: Ginger/Ginger/UserControlsLib/UCComboBox.xaml.cs [hunk 1]
```diff
@@ -23,6 +23,7 @@
 using GingerCore.GeneralLib;
 using System;
 using System.Collections.Generic;
+using System.ComponentModel;
```
Comment (Maheshkale447) [inline Ginger/Ginger/UserControlsLib/UCComboBox.xaml.cs:26]: Remove unwanted usings

**Review by Maheshkale447 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/Actions/UserControls/UCActionDetails.xaml.cs [hunk 2]
```diff
@@ -31,7 +31,7 @@ public partial class UCActionDetails : UserControl
         public UCActionDetails()
         {
             InitializeComponent();
-            this.DataContextChanged += UserControl_DataContextChanged;
+            this.DataContextChanged += UserControl_DataContextChanged;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/Actions/UserControls/UCActionDetails.xaml.cs:34]: not able to apply weak event pattern here

File: Ginger/Ginger/ApplicationModelsLib/POMModels/PomElementsPage.xaml.cs [hunk 3]
```diff
@@ -358,24 +360,26 @@ private void SetElementsGridView()
...
-                xMainElementsGrid.btnAdd.AddHandler(Button.ClickEvent, new RoutedEventHandler(AddUnMappedElementRow));
+                xMainElementsGrid.btnAdd.AddHandler(System.Windows.Controls.Button.ClickEvent, new RoutedEventHandler(AddUnMappedElementRow));
                 xMainElementsGrid.SetbtnDeleteHandler(DeleteUnMappedElementRow);
             }
 
-            xMainElementsGrid.grdMain.PreparingCellForEdit += MainElementsGrid_PreparingCellForEdit;
-            xMainElementsGrid.PasteItemEvent += PasteElementEvent;
+            
+                WeakEventManager<DataGrid, DataGridPreparingCellForEditEventArgs>.AddHandler(source: xMainElementsGrid.grdMain, eventName: nameof(DataGrid.PreparingCellForEdit), handler: MainElementsGrid_PreparingCellForEdit);
+                xMainElementsGrid.PasteItemEvent += PasteElementEvent;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/ApplicationModelsLib/POMModels/PomElementsPage.xaml.cs:377]: Not able to added Weak Event Pattern

File: Ginger/Ginger/ApplicationModelsLib/ModelParams Pages/ModelExpectedValueParamSelectionPage.xaml.cs [hunk 4]
```diff
@@ -62,12 +63,14 @@ public AppModelParameter ShowAsWindow(eWindowShowStyle windowStyle = eWindowShow
             Button selectBtn = new Button();
             selectBtn.Content = "Select";
             selectBtn.Click += new RoutedEventHandler(selectBtn_Click);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/ApplicationModelsLib/ModelParams Pages/ModelExpectedValueParamSelectionPage.xaml.cs:65]: added

File: Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs [hunk 5]
```diff
@@ -526,7 +529,7 @@ private void ActivitiesList_SelectionChanged(object sender, SelectionChangedEven
         {
             if (mActivity != null)
             {
-                mActivity.PropertyChanged -= Activity_PropertyChanged;
+                PropertyChangedEventManager.AddHandler(source: mActivity, handler: Activity_PropertyChanged, propertyName: allProperties);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/AutomatePageLib/NewAutomatePage.xaml.cs:532]: Done

File: Ginger/Ginger/UserControlsLib/UCComboBox.xaml.cs (hunk 1 above)
Comment (prashelke) [reply Ginger/Ginger/UserControlsLib/UCComboBox.xaml.cs]: Done

File: Ginger/Ginger/RunSetPageLib/NewRunSetPage.xaml.cs [hunk 6]
```diff
@@ -743,20 +746,19 @@ void InitRunSetConfigurations()
...
-                xSealightsBuildSessionIDTextBox.ValueTextBox.AddValidationRule(new ValidateEmptyValueWithDependency(mRunSetConfig, nameof(RunSetConfig.SealightsLabId), "Lab ID or Build Session ID must be provided"));
-                xSealightsTestStageTextBox.ValueTextBox.AddValidationRule(new ValidateEmptyValue("Test Stage cannot be empty"));
-            }
+            xSealightsLabIdTextBox.ValueTextBox.AddValidationRule(new ValidateEmptyValueWithDependency(mRunSetConfig, nameof(RunSetConfig.SealightsBuildSessionID), "Lab ID or Build Session ID must be provided"));
+            xSealightsBuildSessionIDTextBox.ValueTextBox.AddValidationRule(new ValidateEmptyValueWithDependency(mRunSetConfig, nameof(RunSetConfig.SealightsLabId), "Lab ID or Build Session ID must be provided"));
+            xSealightsTestStageTextBox.ValueTextBox.AddValidationRule(new ValidateEmptyValue("Test Stage cannot be empty"));
+
             xDefaultTestStageRadioBtn.Checked += XDefaultTestStageRadioBtn_Checked;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/RunSetPageLib/NewRunSetPage.xaml.cs]: Added

File: Ginger/Ginger/RunSetPageLib/NewRunSetPage.xaml.cs [hunk 7]
```diff
@@ -743,20 +746,19 @@ void InitRunSetConfigurations()
             xSealightsBuildSessionIDTextBox.Init(mContext, mRunSetConfig, nameof(RunSetConfig.SealightsBuildSessionID));
 
             // check if fields have been populated (front-end validation)
-            if (!_ignoreValidationRules)
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/RunSetPageLib/NewRunSetPage.xaml.cs]: done

File: Ginger/Ginger/AutomatePageLib/BusinessFlowsAutomatePage.xaml.cs [hunk 8]
```diff
@@ -26,6 +26,8 @@
 using GingerWPF.UserControlsLib;
 using GingerWPF.UserControlsLib.UCTreeView;
 using System;
+using System.ComponentModel;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/AutomatePageLib/BusinessFlowsAutomatePage.xaml.cs]: done

File: Ginger/Ginger/Activities/ActivityPage.xaml.cs [hunk 9]
```diff
@@ -247,8 +262,14 @@ private void XItemsTabs_SelectionChanged(object sender, SelectionChangedEventArg
                         {
                             mActionsPage.ListView.ListTitleVisibility = Visibility.Collapsed;
                         }
-                        mActionsPage.ShiftToActionEditEvent += MActionsPage_ShiftToActionEditEvent;
-                        mActionsPage.ShiftToActionsListEvent += MActionsPage_ShiftToActionsListEvent;
+                        WeakEventManager<ActionsListViewPage, RoutedEventArgs>.AddHandler(source: mActionsPage, eventName: nameof(ActionsListViewPage.ShiftToActionEditEvent), handler: MActionsPage_ShiftToActionEditEvent);
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/Activities/ActivityPage.xaml.cs]: done

File: Ginger/Ginger/Actions/UserControls/UCValueExpression.xaml.cs [hunk 10]
```diff
@@ -22,6 +22,8 @@
 using System;
 using System.Windows;
 using System.Windows.Controls;
+using System.Windows.Controls.Primitives;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/Actions/UserControls/UCValueExpression.xaml.cs]: done

File: Ginger/Ginger/Actions/ActionConversion/SelectActionWzardPage.xaml.cs [hunk 11]
```diff
@@ -23,6 +23,7 @@
 using GingerWPF.WizardLib;
 using System.Collections.Generic;
 using System.Linq;
+using System.Windows;
```
[the remark this answers is not in the document]
Comment (prashelke) [reply Ginger/Ginger/Actions/ActionConversion/SelectActionWzardPage.xaml.cs]: done

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**